waypoint_generator/src/waypoint_generator/waypoint_generator.cpp

```diff
@@ -355,17 +355,37 @@ std::vector<double> WaypointGenerator::get_speed_limits(std::vector<lanelet::Con
   return out;
 }
 
-void WaypointGenerator::new_route_callback()
+std::vector<lanelet::ConstLanelet> WaypointGenerator::findSuccessingLanelets() const
 {
-  ROS_DEBUG_STREAM("Received new route message, processing...");
-
+  std::vector<lanelet::ConstLanelet> out;
   auto shortest_path = _wm->getRoute()->shortestPath();
   ROS_DEBUG_STREAM("Processing " << shortest_path.size() << " lanelets.");
-  std::vector<lanelet::ConstLanelet> tmp;
-  for (lanelet::ConstLanelet l : shortest_path)
+  
+  out.push_back(shortest_path[0]);
+  for (size_t i=1; i<shortest_path.size(); i++)
   {
-    tmp.push_back(l);
+    lanelet::ConstLanelet l = shortest_path[i];
+    auto following = _wm->getRoute()->followingRelations(shortest_path[i-1]);
+    if (following[0].lanelet.id()==l.id() && following[0].relationType == lanelet::routing::RelationType::Successor){
+      out.push_back(l);
+    }
+    else{
```

Comment:
Rather than logging an error. I think we should probably throw an exception. The curvature calculation will not work in this case, so we shouldn't let the vehicle drive this route. If you throw an exception you can also use ASSERT_THROW in your unit test to validate the event.